File: internal/integrations/argocd/actions.go
package argocd

import (
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/roee-hersh/hallpass/internal/catalog"
	"github.com/roee-hersh/hallpass/internal/integrations/argocd/rbac"
)

// actRollback is resolved at check time to sync or rollback depending on
// server.rbac.rollback.enforce.enable.
const actRollback = "\x00rollback"

type action struct {
	name, desc string
	res, act   string
}

var actionList = []action{
	{"app.get", "view an application", rbac.ResourceApplications, rbac.ActionGet},
	{"app.create", "create an application", rbac.ResourceApplications, rbac.ActionCreate},
	{"app.update", "update an application", rbac.ResourceApplications, rbac.ActionUpdate},
	{"app.delete", "delete an application", rbac.ResourceApplications, rbac.ActionDelete},
	{"app.sync", "sync an application", rbac.ResourceApplications, rbac.ActionSync},
	{"app.rollback", "roll an application back (sync, or rollback when server.rbac.rollback.enforce.enable is on)", rbac.ResourceApplications, actRollback},
	{"app.override", "override application parameters", rbac.ResourceApplications, rbac.ActionOverride},
	{"logs.get", "read pod logs of an application", rbac.ResourceLogs, rbac.ActionGet},
	{"exec.create", "exec into a pod of an application", rbac.ResourceExec, rbac.ActionCreate},
	{"appset.get", "view an ApplicationSet", rbac.ResourceApplicationSets, rbac.ActionGet},
	{"appset.create", "create an ApplicationSet", rbac.ResourceApplicationSets, rbac.ActionCreate},
	{"appset.update", "update an ApplicationSet", rbac.ResourceApplicationSets, rbac.ActionUpdate},
	{"appset.delete", "delete an ApplicationSet", rbac.ResourceApplicationSets, rbac.ActionDelete},
	{"project.get", "view a project", rbac.ResourceProjects, rbac.ActionGet},
	{"project.create", "create a project", rbac.ResourceProjects, rbac.ActionCreate},
	{"project.update", "update a project", rbac.ResourceProjects, rbac.ActionUpdate},
	{"project.delete", "delete a project", rbac.ResourceProjects, rbac.ActionDelete},
	{"cluster.get", "view a cluster", rbac.ResourceClusters, rbac.ActionGet},
	{"cluster.create", "add a cluster", rbac.ResourceClusters, rbac.ActionCreate},
	{"cluster.update", "update a cluster", rbac.ResourceClusters, rbac.ActionUpdate},
	{"cluster.delete", "remove a cluster", rbac.ResourceClusters, rbac.ActionDelete},
	{"repo.get", "view a repository", rbac.ResourceRepositories, rbac.ActionGet},
	{"repo.create", "add a repository", rbac.ResourceRepositories, rbac.ActionCreate},
	{"repo.update", "update a repository", rbac.ResourceRepositories, rbac.ActionUpdate},
	{"repo.delete", "remove a repository", rbac.ResourceRepositories, rbac.ActionDelete},
	{"writerepo.get", "view a write repository", rbac.ResourceWriteRepositories, rbac.ActionGet},
	{"writerepo.create", "add a write repository", rbac.ResourceWriteRepositories, rbac.ActionCreate},
	{"writerepo.update", "update a write repository", rbac.ResourceWriteRepositories, rbac.ActionUpdate},
	{"writerepo.delete", "remove a write repository", rbac.ResourceWriteRepositories, rbac.ActionDelete},
	{"certificate.get", "view a certificate", rbac.ResourceCertificates, rbac.ActionGet},
	{"certificate.create", "add a certificate", rbac.ResourceCertificates, rbac.ActionCreate},
	{"certificate.update", "update a certificate", rbac.ResourceCertificates, rbac.ActionUpdate},
	{"certificate.delete", "remove a certificate", rbac.ResourceCertificates, rbac.ActionDelete},
	{"account.get", "view an account", rbac.ResourceAccounts, rbac.ActionGet},
	{"account.update", "update an account", rbac.ResourceAccounts, rbac.ActionUpdate},
	{"gpgkey.get", "view a GPG key", rbac.ResourceGPGKeys, rbac.ActionGet},
	{"gpgkey.create", "add a GPG key", rbac.ResourceGPGKeys, rbac.ActionCreate},
	{"gpgkey.delete", "remove a GPG key", rbac.ResourceGPGKeys, rbac.ActionDelete},
	{"extension.invoke", "invoke an extension", rbac.ResourceExtensions, rbac.ActionInvoke},
}

var actions = func() map[string]action {
	m := map[string]action{}
	for _, a := range actionList {
		m[a.name] = a
	}
	return m
}()

// resourceTypes maps hallpass resource types to Argo CD resources.
var resourceTypes = map[string]string{
	"applications":       rbac.ResourceApplications,
	"applicationsets":    rbac.ResourceApplicationSets,
	"logs":               rbac.ResourceLogs,
	"exec":               rbac.ResourceExec,
	"projects":           rbac.ResourceProjects,
	"clusters":           rbac.ResourceClusters,
	"repositories":       rbac.ResourceRepositories,
	"write_repositories": rbac.ResourceWriteRepositories,
	"certificates":       rbac.ResourceCertificates,
	"accounts":           rbac.ResourceAccounts,
	"gpgkeys":            rbac.ResourceGPGKeys,
	"extensions":         rbac.ResourceExtensions,
}

// request is what goes into the enforcer.
type request struct {
	res, act, obj string
	// fineGrained update/delete carry the top-level verb for v2 inheritance.
	fineGrained bool
	topAct      string
}

var (
	segmentRe = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]{0,252}$`)
	// objRe bounds the object string. It is matched as a value, never as a
	// pattern, so it only needs to be printable.
	objRe = regexp.MustCompile(`^[^\x00-\x20\x7f]{1,512}$`)
)

// parsePattern recognises app.action/<g>/<k>/<a>, app.update/<g>/<k>/<ns>/<n>
// and app.delete/<g>/<k>/<ns>/<n>.
func parsePattern(name string) (request, bool) {
	prefix, rest, ok := strings.Cut(name, "/")
	if !ok {
		return request{}, false
	}
	parts := strings.Split(rest, "/")
	for _, p := range parts {
		if p != "" && !segmentRe.MatchString(p) {
			return request{}, false
		}
	}
	switch prefix {
	case "app.action":
		if len(parts) != 3 || parts[1] == "" || parts[2] == "" {
			return request{}, false
		}
		return request{res: rbac.ResourceApplications, act: rbac.ActionAction + "/" + rest}, true
	case "app.update", "app.delete":
		if len(parts) != 4 || parts[1] == "" || parts[3] == "" {
			return request{}, false
		}
		verb := strings.TrimPrefix(prefix, "app.")
		return request{res: rbac.ResourceApplications, act: verb + "/" + rest, fineGrained: true, topAct: verb}, true
	}
	return request{}, false
}

// buildRequest combines the action and the resource.
func buildRequest(actionName string, res catalog.Resource) (request, error) {
	var req request
	if a, ok := actions[actionName]; ok {
		req = request{res: a.res, act: a.act}
	} else if p, ok := parsePattern(actionName); ok {
		req = p
	} else {
		return req, fmt.Errorf("unknown action %q", actionName)
	}
	argoRes, ok := resourceTypes[res.Type]
	if !ok {
		return req, fmt.Errorf("resource type %q; use one of applications, applicationsets, logs, exec, projects, clusters, repositories, write_repositories, certificates, accounts, gpgkeys, extensions", res.Type)
	}
	// logs and exec take an applications resource too, since the object is the same.
	if argoRes != req.res && !((req.res == rbac.ResourceLogs || req.res == rbac.ResourceExec) && argoRes == rbac.ResourceApplications) {
		return req, fmt.Errorf("action %s applies to %s, but the resource is %s", actionName, req.res, res.Type)
	}
	if res.ID == "" {
		return req, errors.New("resource needs an id, e.g. applications:<project>/<name>")
	}
	if !objRe.MatchString(res.ID) {
		return req, errors.New("resource id contains whitespace or control characters")
	}
	if rbac.ProjectScoped[req.res] && req.res != rbac.ResourceClusters && req.res != rbac.ResourceRepositories {
		if !strings.Contains(res.ID, "/") {
			return req, fmt.Errorf("%s resources are <project>/<name>", res.Type)
		}
	}
	req.obj = res.ID
	return req, nil
}

File: hallpass-py/src/hallpass/integrations/argocd/actions.py
"""The argocd actions and how a check becomes an enforcer request."""

from __future__ import annotations

import re
from dataclasses import dataclass

from hallpass.core.catalog import Resource
from hallpass.core.errors import go_quote

from . import rbac

__all__ = ["ACTIONS", "ACTION_LIST", "ACT_ROLLBACK", "OBJ_RE", "RESOURCE_TYPES", "ActionDef", "EnforceRequest", "build_request", "parse_pattern"]

# Resolved at check time to sync or rollback depending on
# server.rbac.rollback.enforce.enable.
ACT_ROLLBACK = "\x00rollback"


@dataclass(frozen=True)
class ActionDef:
    name: str
    desc: str
    res: str
    act: str


ACTION_LIST: tuple[ActionDef, ...] = (
    ActionDef("app.get", "view an application", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_GET),
    ActionDef("app.create", "create an application", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_CREATE),
    ActionDef("app.update", "update an application", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_UPDATE),
    ActionDef("app.delete", "delete an application", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_DELETE),
    ActionDef("app.sync", "sync an application", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_SYNC),
    ActionDef(
        "app.rollback",
        "roll an application back (sync, or rollback when server.rbac.rollback.enforce.enable is on)",
        rbac.RESOURCE_APPLICATIONS,
        ACT_ROLLBACK,
    ),
    ActionDef("app.override", "override application parameters", rbac.RESOURCE_APPLICATIONS, rbac.ACTION_OVERRIDE),
    ActionDef("logs.get", "read pod logs of an application", rbac.RESOURCE_LOGS, rbac.ACTION_GET),
    ActionDef("exec.create", "exec into a pod of an application", rbac.RESOURCE_EXEC, rbac.ACTION_CREATE),
    ActionDef("appset.get", "view an ApplicationSet", rbac.RESOURCE_APPLICATION_SETS, rbac.ACTION_GET),
    ActionDef("appset.create", "create an ApplicationSet", rbac.RESOURCE_APPLICATION_SETS, rbac.ACTION_CREATE),
    ActionDef("appset.update", "update an ApplicationSet", rbac.RESOURCE_APPLICATION_SETS, rbac.ACTION_UPDATE),
    ActionDef("appset.delete", "delete an ApplicationSet", rbac.RESOURCE_APPLICATION_SETS, rbac.ACTION_DELETE),
    ActionDef("project.get", "view a project", rbac.RESOURCE_PROJECTS, rbac.ACTION_GET),
    ActionDef("project.create", "create a project", rbac.RESOURCE_PROJECTS, rbac.ACTION_CREATE),
    ActionDef("project.update", "update a project", rbac.RESOURCE_PROJECTS, rbac.ACTION_UPDATE),
    ActionDef("project.delete", "delete a project", rbac.RESOURCE_PROJECTS, rbac.ACTION_DELETE),
    ActionDef("cluster.get", "view a cluster", rbac.RESOURCE_CLUSTERS, rbac.ACTION_GET),
    ActionDef("cluster.create", "add a cluster", rbac.RESOURCE_CLUSTERS, rbac.ACTION_CREATE),
    ActionDef("cluster.update", "update a cluster", rbac.RESOURCE_CLUSTERS, rbac.ACTION_UPDATE),
    ActionDef("cluster.delete", "remove a cluster", rbac.RESOURCE_CLUSTERS, rbac.ACTION_DELETE),
    ActionDef("repo.get", "view a repository", rbac.RESOURCE_REPOSITORIES, rbac.ACTION_GET),
    ActionDef("repo.create", "add a repository", rbac.RESOURCE_REPOSITORIES, rbac.ACTION_CREATE),
    ActionDef("repo.update", "update a repository", rbac.RESOURCE_REPOSITORIES, rbac.ACTION_UPDATE),
    ActionDef("repo.delete", "remove a repository", rbac.RESOURCE_REPOSITORIES, rbac.ACTION_DELETE),
    ActionDef("writerepo.get", "view a write repository", rbac.RESOURCE_WRITE_REPOSITORIES, rbac.ACTION_GET),
    ActionDef("writerepo.create", "add a write repository", rbac.RESOURCE_WRITE_REPOSITORIES, rbac.ACTION_CREATE),
    ActionDef("writerepo.update", "update a write repository", rbac.RESOURCE_WRITE_REPOSITORIES, rbac.ACTION_UPDATE),
    ActionDef("writerepo.delete", "remove a write repository", rbac.RESOURCE_WRITE_REPOSITORIES, rbac.ACTION_DELETE),
    ActionDef("certificate.get", "view a certificate", rbac.RESOURCE_CERTIFICATES, rbac.ACTION_GET),
    ActionDef("certificate.create", "add a certificate", rbac.RESOURCE_CERTIFICATES, rbac.ACTION_CREATE),
    ActionDef("certificate.update", "update a certificate", rbac.RESOURCE_CERTIFICATES, rbac.ACTION_UPDATE),
    ActionDef("certificate.delete", "remove a certificate", rbac.RESOURCE_CERTIFICATES, rbac.ACTION_DELETE),
    ActionDef("account.get", "view an account", rbac.RESOURCE_ACCOUNTS, rbac.ACTION_GET),
    ActionDef("account.update", "update an account", rbac.RESOURCE_ACCOUNTS, rbac.ACTION_UPDATE),
    ActionDef("gpgkey.get", "view a GPG key", rbac.RESOURCE_GPG_KEYS, rbac.ACTION_GET),
    ActionDef("gpgkey.create", "add a GPG key", rbac.RESOURCE_GPG_KEYS, rbac.ACTION_CREATE),
    ActionDef("gpgkey.delete", "remove a GPG key", rbac.RESOURCE_GPG_KEYS, rbac.ACTION_DELETE),
    ActionDef("extension.invoke", "invoke an extension", rbac.RESOURCE_EXTENSIONS, rbac.ACTION_INVOKE),
)

ACTIONS: dict[str, ActionDef] = {a.name: a for a in ACTION_LIST}

# Maps hallpass resource types to Argo CD resources.
RESOURCE_TYPES: dict[str, str] = {
    "applications": rbac.RESOURCE_APPLICATIONS,
    "applicationsets": rbac.RESOURCE_APPLICATION_SETS,
    "logs": rbac.RESOURCE_LOGS,
    "exec": rbac.RESOURCE_EXEC,
    "projects": rbac.RESOURCE_PROJECTS,
    "clusters": rbac.RESOURCE_CLUSTERS,
    "repositories": rbac.RESOURCE_REPOSITORIES,
    "write_repositories": rbac.RESOURCE_WRITE_REPOSITORIES,
    "certificates": rbac.RESOURCE_CERTIFICATES,
    "accounts": rbac.RESOURCE_ACCOUNTS,
    "gpgkeys": rbac.RESOURCE_GPG_KEYS,
    "extensions": rbac.RESOURCE_EXTENSIONS,
}


@dataclass
class EnforceRequest:
    """What goes into the enforcer."""

    res: str
    act: str
    obj: str = ""
    # Fine-grained update/delete carry the top-level verb for v2 inheritance.
    fine_grained: bool = False
    top_act: str = ""


SEGMENT_RE = re.compile(r"[A-Za-z0-9][A-Za-z0-9._-]{0,252}")
# Bounds the object string. It is matched as a value, never as a pattern,
# so it only needs to be printable. Always used with fullmatch.
OBJ_RE = re.compile(r"[^\x00-\x20\x7f]{1,512}")


def parse_pattern(name: str) -> EnforceRequest | None:
    """Recognise app.action/<g>/<k>/<a>, app.update/<g>/<k>/<ns>/<n> and
    app.delete/<g>/<k>/<ns>/<n>."""
    prefix, sep, rest = name.partition("/")
    if not sep:
        return None
    parts = rest.split("/")
    for p in parts:
        if p != "" and SEGMENT_RE.fullmatch(p) is None:
            return None
    if prefix == "app.action":
        if len(parts) != 3 or parts[1] == "" or parts[2] == "":
            return None
        return EnforceRequest(res=rbac.RESOURCE_APPLICATIONS, act=rbac.ACTION_ACTION + "/" + rest)
    if prefix in ("app.update", "app.delete"):
        if len(parts) != 4 or parts[1] == "" or parts[3] == "":
            return None
        verb = prefix[len("app.") :]
        return EnforceRequest(res=rbac.RESOURCE_APPLICATIONS, act=verb + "/" + rest, fine_grained=True, top_act=verb)
    return None


def build_request(action_name: str, res: Resource) -> EnforceRequest:
    """Combine the action and the resource. Raises ValueError."""
    a = ACTIONS.get(action_name)
    if a is not None:
        req = EnforceRequest(res=a.res, act=a.act)
    else:
        p = parse_pattern(action_name)
        if p is None:
            raise ValueError(f"unknown action {go_quote(action_name)}")
        req = p
    argo_res = RESOURCE_TYPES.get(res.type)
    if argo_res is None:
        raise ValueError(
            f"resource type {go_quote(res.type)}; use one of applications, applicationsets, logs, exec, projects, clusters, repositories, "
            "write_repositories, certificates, accounts, gpgkeys, extensions"
        )
    # logs and exec take an applications resource too, since the object is the same.
    if argo_res != req.res and not (req.res in (rbac.RESOURCE_LOGS, rbac.RESOURCE_EXEC) and argo_res == rbac.RESOURCE_APPLICATIONS):
        raise ValueError(f"action {action_name} applies to {req.res}, but the resource is {res.type}")
    if res.id == "":
        raise ValueError("resource needs an id, e.g. applications:<project>/<name>")
    if OBJ_RE.fullmatch(res.id) is None:
        raise ValueError("resource id contains whitespace or control characters")
    if req.res in rbac.PROJECT_SCOPED and req.res not in (rbac.RESOURCE_CLUSTERS, rbac.RESOURCE_REPOSITORIES) and "/" not in res.id:
        raise ValueError(f"{res.type} resources are <project>/<name>")
    req.obj = res.id
    return req
